pilot/pkg/autoregistration/controller.go

```diff
@@ -229,15 +229,26 @@ func (c *Controller) RegisterWorkload(proxy *model.Proxy, conTime time.Time) err
 }
 
 func (c *Controller) registerWorkload(entryName string, proxy *model.Proxy, conTime time.Time) error {
+	groupCfg := c.store.Get(gvk.WorkloadGroup, proxy.Metadata.AutoRegisterGroup, proxy.Metadata.Namespace)
+	if groupCfg == nil {
+		autoRegistrationErrors.Increment()
+		return grpcstatus.Errorf(codes.FailedPrecondition, "auto-registration WorkloadEntry of %v failed: cannot find WorkloadGroup %s/%s",
+			proxy.ID, proxy.Metadata.Namespace, proxy.Metadata.AutoRegisterGroup)
+	}
+
 	wle := c.store.Get(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace)
 	if wle != nil {
 		lastConTime, _ := time.Parse(timeFormat, wle.Annotations[ConnectedAtAnnotation])
 		// the proxy has reconnected to another pilot, not belong to this one.
 		if conTime.Before(lastConTime) {
 			return nil
 		}
-		// Try to patch, if it fails then try to create
 		_, err := c.store.Patch(*wle, func(cfg config.Config) (config.Config, kubetypes.PatchType) {
+			nextEntry := workloadEntryFromGroup(entryName, proxy, groupCfg)
+			// copy things as needed since cfg will have things like resource version and uid set by k8s
```

Comment:
resource version and UID are not in labels/annotations though?

Note labels are duplicated in `spec` as well, is that a concern?

Reply:
I think its a valid concern btw since a lot of things add labels and annotations dynamically. But its more a sign we should use SSA IMO. https://github.com/istio/istio/blob/c2d6094a5b33c76c81bb969ca08c1181147e692a/pilot/pkg/config/kube/gateway/deploymentcontroller.go#L67 goes into this quite a bit

Reply:
I think the comment is confusing. I'm just saying we can't return `nextEntry` here because that wouldn't include some critical fields that `cfg` has. 

I'll look into SSA. IIRC that's what we tried to use first but there were properties of Patch that made it handle races better. Maybe it was library support.

Reply:
So in the meantime should we:
* merge this as is
* use the yaml+Patch based approach you linked to?

Reply:
I am only like 75% confident in everything below, so we should probably test and see what works. Anything that works SGTM:

I think if we merge as is then you just cannot remove fields. That generally only makes sense for labels and annotations anyways though. I guess you could remove `network` or something but its a bit quirky.

If we can get raw yaml (json works too) that is correct that works as well.

I know my comments about generally https://github.com/istio/istio/pull/44741 not working, but I think it probably works fine for or WE use cases. That is, we can just do our patch exactly as-is and use SSA apply and it should be fine (for Spec - ObjectMeta needs to be handled.

Note if we use SSA we won't want to do mergeLabels since k8s will handle that.

---

If we don't use SSA, is mergeLabels correct anyways? will the patch merge labels or overwrite the entire thing?

Reply:
It doesn't overwrite by patching at the `labels` key granularity, but by generating a diff and putting `null` at keys that it doesn't see in the new version. 

https://gist.github.com/stevenctl/d064c8afd04e8246c2b54fa008281b6b

This results in 

```
{"metadata":{"labels":{"newlabel":"newvalue","oldlabel":null}},"spec":{"labels":{"newlabel":"newvalue","oldlabel":null}}}
```

setConnectMeta on top of the original worked fine

---

If we _do_ use SSA, my hope is that only including the labels we care about will update those and the fieldmanager logic will leave labels managed by other things alone

On the note of mirrored labels in spec, I'd imagine other controllers mostly don't respect the labels in Spec? 

I'd also hope code that reads the resource takes actual labels as more of a source of truth where available. The spec labels seems mostly useful for the ServiceEntry usecase I think it would be a bit better if they were discouraged for standalone WLE.

Reply:
In short, this does work.

Reply:
One idea to work around this:

Consider the generated .Spec.Labels as the stuff we manage. It should be safe to remove labels that disappear there. 

Other labels should be preserved. 

WDYT?

Reply:
Only concern is it relies on use seeing the diff of old vs new on the WG, right? Not sure we reliably get that

Reply:
```
---
# old server value
meta:
  labels:
    somecontroller.k8s.io: bar
    topology.istio.io/network: remote
    workloadgrouplabel1: example
    workloadgrouplabel2: example
spec:
  labels:
    topology.istio.io/network: remote
    workloadgrouplabel1: example
    workloadgrouplabel2: example
---
# newly generated from WorkloadGroup + connection metadata
meta:
  labels:
    topology.istio.io/network: remote
    workloadgrouplabel1: changed
spec:
  labels:
    topology.istio.io/network: remote
    workloadgrouplabel1: changed
---
# merged output
meta:
  labels:
    somecontroller.k8s.io: bar
    topology.istio.io/network: remote
    workloadgrouplabel1: changed
spec:
  labels:
    topology.istio.io/network: remote
    workloadgrouplabel1: changed
```

*  spec,labels are always derived from current workloadgroup and connection metadata
*  metadata.labels are:
    * (server metadata.labels - server spec.labels) + generated spec.labels